KSP-CKAN/CKAN
# Fix auto-remove during upgrade

## Problem

In https://github.com/KSP-RO/RP-1/commit/b9f085575debb6be46dfeed3f722f3110ce0d046, `RP-1`'s dependency on `MagiCore` was removed from its metanetkan, and this change first reached users with the indexing of `RP-1` v2.12 in https://github.com/KSP-CKAN/CKAN-meta/commit/98a346180f649c277e026a18cd69d0c3b5055ac8. When they upgraded, `RP-1-ExpressInstall` and all of its auto-installed dependencies were uninstalled (which should not have happened!). The dependency was subsequently restored pending a CKAN fix.

### How to reproduce for testing

1. Use a fresh game instance of KSP1 1.12.5
2. In CKAN, go to the settings and replace the default repo with this URL to simulate using CKAN before the problematic upgrade (name doesn't matter):
   https://github.com/KSP-CKAN/CKAN-meta/archive/0451525c2c9cefd16914032a2ca38d69c9e3f21c.tar.gz
3. Click Refresh to update metadata (will happen automatically if using a current dev build)
4. Install `RP-1-ExpressInstall`
5. When prompted to choose a version of `RP-1`, choose the 2.x series version
   (the other choices you make don't matter, issue will happen regardless)
6. Note that with the current dev build, there will be an extremely long delay while it computes the recommendations on the extremely long `RP-1` dependency list because #3892 was merged, ~~KSP-RO/RP-1#2233 was reverted, and KSP-RO/RP-1-ExpressInstall-Graphics#4 and KSP-RO/RealismOverhaul#2913 have not yet been merged. I have sent a reminder about this.~~
   Those pull requests have now been merged, however I forgot that that will not help us here as we are intentionally reaching back into the archive to use old versions of the metadata for testing. But at least now it will not be as slow after the next client release.
7. Now go back to the CKAN settings and replace your repo with this URL to simulate using CKAN right after the release of RP-1 v2.12:
   https://github.com/KSP-CKAN/CKAN-meta/archive/98a346180f649c277e026a18cd69d0c3b5055ac8.tar.gz
8. Click Refresh to update metadata (will happen automatically if using a current dev build).
   You should see four mods ready to upgrade.
9. Choose to upgrade those mods and click Apply
   ![image](https://github.com/KSP-CKAN/CKAN/assets/1559108/4243480b-0177-4035-8fa8-5841f9a813e4)
10. This is where `RP-1-ExpressInstall` (and many other mods) would have been removed previously. Instead you should see this with this PR's changes:
   ![image](https://github.com/KSP-CKAN/CKAN/assets/1559108/f8b264c6-32eb-4c63-a615-e69060805ef8)
11. Finish upgrading and confirm that `MagiCore` is gone and `RP-1-ExpressInstall` is still there

## Cause

- Upgrading RP-1 triggers auto-removal of MagiCore, as it should
- GUI always calls `ModuleInstaller.UninstallList` before `ModuleInstaller.InstallList` before `ModuleInstaller.Upgrade`:
  https://github.com/KSP-CKAN/CKAN/blob/341adf88886d6d86626bca0073421cf576af5081/GUI/Main/MainInstall.cs#L169-L184
  ... so the first change that happens is removal of MagiCore
- The CKAN core code performing that removal doesn't know anything about the pending upgrade, it just knows it's been asked to remove MagiCore and that the currently installed version of RP-1 depends on MagiCore
- When that code removes MagiCore, it also removes all mods that depend on it, including the currently installed (old) version of RP-1. This is mildly surprising and not ideal, but not necessarily a serious problem yet.
- However, that dependency search is **transitive**, so everything depending on RP-1 is also  removed, including RP-1 Express! **This is when things definitely are going wrong**.
- The removal of RP-1 Express also triggers the automatic removal of any mods marked as auto-installed that only RP-1 Express depended on. This is desirable in a narrow sense, but in the context of removing a mod that we don't really want to remove, it just makes a bigger mess.
- When the Remove step finishes, essentially everything is uninstalled
- Since we're not installing anything, the InstallList step does nothing
- Finally the Upgrade step happens, which causes all the "Update" mods in the changeset to be reinstalled:
  ![image](https://github.com/KSP-CKAN/CKAN/assets/1559108/04a875a5-6944-4cd2-b1f9-b400fde9698e)
  ... so RP-1, RealismOverhaul, RealFuels, and KSPCommunityFixes.
- But RP-1 Express was never in the changeset, so it doesn't get reinstalled!

## Changes

- Now `ModuleInstaller.Upgrade` handles auto-removal internally
  (`ModuleInstaller.UninstallList` already handled auto-removal internally)
  - Tests are added for auto-removal
- Now GUI doesn't pass auto-removals to `ModuleInstaller.UninstallList` unless the changeset consists only of installs and auto-removals, because we want them to be handled by `Upgrade` and/or `UninstallList` whenever possible
- While working on this, I discovered that the refactor of the DLL scanning code from #3904 throws if you have multiple DLLs with the same filename (which the KSP-RO team is currently doing unintentionally by shipping `System.dll`, `System.Core.dll`, and `System.Xml.dll` in several of their ZIPs). Now this is fixed courtesy of `GroupBy`.
- While working on this, I discovered that `RP-1-ExpressInstall` has enough tags to wrap the display in GUI, but the panel containing them doesn't resize to fit. Now it does.
- While working on this, I noticed that if you select a very large changeset, it takes a long time to calculate the recommendations, during which CKAN shows the changeset tab and is responsive but doesn't look busy. Now the hourglass cursor appears during this to show that it is doing something.
- The right-click reinstall option now uses `Upgrade` instead of cobbling together its own custom changeset of uninstalls and installs, to take advantage of this smarter handling of auto-removal for that use case (fixes the part of #3912 where it's unnecessarily trying to uninstall Parallax-StockTextures). When this option is used, the changeset will say "Re-install (user requested)" in the English locale.

Fixes #3911.
